protocols/virtual-fund/virtual-fund.go

```diff
@@ -0,0 +1,369 @@
+	"github.com/statechannels/go-nitro/channel/state"
+	"github.com/statechannels/go-nitro/channel/state/outcome"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/types"
+)
+
+const (
+	WaitingForCompletePrefund  protocols.WaitingFor = "WaitingForCompletePrefund"  // Round 1
+	WaitingForCompleteFunding  protocols.WaitingFor = "WaitingForCompleteFunding"  // Round 2
+	WaitingForCompletePostFund protocols.WaitingFor = "WaitingForCompletePostFund" // Round 3
+	WaitingForNothing          protocols.WaitingFor = "WaitingForNothing"          // Finished
+)
+
+var NoSideEffects = protocols.SideEffects{}
+
+// errors
+var ErrNotApproved = errors.New("objective not approved")
+
+type Connection struct {
+	Channel            channel.Channel
+	ExpectedGuarantees map[types.Address]outcome.Allocation
+}
+
+// VirtualFundObjective is a cache of data computed by reading from the store. It stores (potentially) infinite data.
+type VirtualFundObjective struct {
+	Status protocols.ObjectiveStatus
+	V      *channel.Channel
+
+	ToMyLeft  *Connection
+	ToMyRight *Connection
+
+	n      uint // number of ledger channels (num_hops + 1)
+	MyRole uint // index in the virtual funding protocol. 0 for Alice, n+1 for Bob. Otherwise, one of the intermediaries.
+
+	a0 types.Funds // Initial balance for Alice
+	b0 types.Funds // Initial balance for Bob
+
+	requestedLedgerUpdates bool // records that the ledger update side effects were previously generated (they may not have been executed yet)
+}
+
+////////////////////////////////////////////////
+// Public methods on the VirtualFundObjective //
+////////////////////////////////////////////////
+
+// New initiates a VirtualFundObjective.
+func New(
+	initialStateOfV state.State,
+	myAddress types.Address,
+	n uint, // number of ledger channels (num_hops + 1)
+	myRole uint,
+	ledgerChannelToMyLeft channel.Channel,
+	ledgerChannelToMyRight channel.Channel,
+) (VirtualFundObjective, error) {
+	// role and ledger-channel checks
+	if myRole > n+1 {
+		return VirtualFundObjective{}, fmt.Errorf(`invalid role <%d> specified in %d-hop virtual-fund objective`,
+			myRole, n-1)
+	}
+
+	switch myRole {
+	case 0: // Alice
+		if !ledgerChannelToMyRight.IsTwoPartyLedger {
+			return VirtualFundObjective{}, errors.New(`alice's right-channel is not a two-party ledger channel`)
+		}
+	case n + 1: // Bob
+		if !ledgerChannelToMyLeft.IsTwoPartyLedger {
+			return VirtualFundObjective{}, errors.New(`bob's left-channel is not a two-party ledger channel`)
+		}
+	default: // Intermediary
+		if !(ledgerChannelToMyLeft.IsTwoPartyLedger && ledgerChannelToMyRight.IsTwoPartyLedger) {
+			return VirtualFundObjective{}, errors.New(`supplied channels are not two party ledger channels`)
+		}
+	}
+
+	var init VirtualFundObjective
+
+	// Initialize virtual channel
+	v, err := channel.New(initialStateOfV, false, myRole, types.Destination{}, types.Destination{})
+	if err != nil {
+		return init, err
+
+	}
```

Comment:
Explicitly empty returned value (matching other err returns)

```suggestion
		return VirtualFundObjective{}, err
	}

```